test_netcdf_timeaverage.jl

```diff
@@ -0,0 +1,213 @@
+                                timestepper = :RungeKutta3,
+                                tracers = (:c1, :c2),
+                                forcing = (c1=c1_forcing, c2=c2_forcing))
+
+
+    Nx, Ny, Nz = size(grid)
+    xs, ys, zs = nodes(model.tracers.c1)
+
+    c̄1(z, t) = 1 / (Nx * Ny) * sum(exp(-λ1(x, y, z) * t) for x in xs for y in ys)
+    c̄2(z, t) = 1 / (Nx * Ny) * sum(exp(-λ2(x, y, z) * t) for x in xs for y in ys)
+
+    rtol = 1e-5 # need custom rtol for isapprox because roundoff errors accumulate (?)
+
+    # Compute time averages...
+    c̄1(ts) = 1/length(ts) * sum(c̄1.(zs, t) for t in ts)
+    c̄2(ts) = 1/length(ts) * sum(c̄2.(zs, t) for t in ts)
+
+    #####
+    ##### Test strided windowed time average against analytic solution
+    ##### for *single* NetCDF output
+    #####
+    single_time_average_nc_filepath = "single_decay_windowed_time_average_test.nc"
+    single_ds = NCDataset(single_time_average_nc_filepath)
+
+    attribute_names = ("schedule", "interval", "output time interval",
+                       "time_averaging_window", "time averaging window",
+                       "time_averaging_stride", "time averaging stride")
+
+    for name in attribute_names
+        @test haskey(single_ds.attrib, name) && !isnothing(single_ds.attrib[name])
+    end
+
+    window_size = window_nΔt
+    window = window_size*Δt
+    # @info "    Testing time-averaging of a single NetCDF output [$(typeof(arch))]..."
+
+    for (n, t) in enumerate(single_ds["time"][2:end])
+        averaging_times = [t - n*Δt for n in 0:stride:window_size-1 if t - n*Δt >= 0]
+        # @info n,t,averaging_times, c̄1(averaging_times), single_ds["c1"][:, n+1], c̄1(averaging_times)./single_ds["c1"][:, n+1]
+        @test all(isapprox.(single_ds["c1"][:, n+1], c̄1(averaging_times), rtol=rtol))
+    end
+
+    time = single_ds["time"][:]
+    data_plot = single_ds["c1"][1:4, :]
+    c̄1_timeaverage = zeros(4,length(time[1:end]))
+    for (n, t) in enumerate(time[1:end])
+        averaging_times = [t - n*Δt for n in 0:stride:window_size-1 if t - n*Δt >= 0]
+        # @info n,t,averaging_times, c̄1(averaging_times)
+        c̄1_timeaverage[:,n] = c̄1(averaging_times)
+        # @test all(isapprox.(single_ds["c1"][:, n+1], c̄1(averaging_times), rtol=rtol))
+    end
+
+    # Plot each of the four lines
+    pl = plot()
+    plot!(time, data_plot[1, :], label="1", color=:blue, legend=:topright)
+    plot!(time, data_plot[2, :], label="2", color=:red)
+    plot!(time, data_plot[3, :], label="3", color=:orange)
+    plot!(time, data_plot[4, :], label="4", color=:green)
+    
+    plot!(time[1:end],c̄1_timeaverage[1,:], color=:black, linestyle=:dash, label="1-analytic")
+    plot!(time[1:end],c̄1_timeaverage[2,:], color=:black, linestyle=:dash, label="2-analytic")
+    plot!(time[1:end],c̄1_timeaverage[3,:], color=:black, linestyle=:dash, label="3-analytic")
+    plot!(time[1:end],c̄1_timeaverage[4,:], color=:black, linestyle=:dash, label="4-analytic")
```

Comment:
Hmm, this file is named `test_netcdf_timeaverage.jl`, but since you're plotting things, this is looking more like a `/validation` file than a `/test` file. You could separate this into two files: one automated, no-plotting file, which would be run in our `/test` suite, and another one with the plots, which can live in `/validation`. Alternatively you can just to the `/test` script, which I think is necessary.

Reply:
I agree, this file is more like a validation file that gives me more flexibility to explore. I believe the original test file works well enough for testing purposes.

Reply:
If you want to keep plotting code around, move it to `validation` in the appropriate place. But its best to figure out how to write the tests to encapsulate as much information as possible that would otherwise be gleaned from plotting